File: c#/Sodium/src/Sodium/Node.cs
using System;
using System.Collections.Generic;

namespace Sodium
{
  public class Node : IComparable<Node>
  {
    private long rank;
    public static readonly Node Null = new Node(long.MaxValue);

    public Node(long rank)
    {
      this.rank = rank;
    }
    public int CompareTo(Node other)
    {
      if (rank < other.rank) return -1;
      if (rank > other.rank) return 1;
      return 0;
    }

    private readonly HashSet<Node> listeners = new HashSet<Node>();

    public bool LinkTo(Node target)
    {
      if (target == Null)
        return false;
      var changed = target.EnsureBiggerThan(rank, new HashSet<Node>());
      listeners.Add(target);
      return changed;
    }

    public void UnlinkTo(Node target)
    {
      if (target == Null)
        return;
      listeners.Remove(target);
    }

    bool EnsureBiggerThan(long limit, HashSet<Node> visited)
    {
      if (rank > limit || visited.Contains(this))
        return false;
      visited.Add(this);
      rank = limit + 1;
      foreach (var listener in listeners)
      {
        listener.EnsureBiggerThan(rank, visited);
      }
      return true;
    }
  }
}

File: java/sodium/src/sodium/Node.java
package sodium;

import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.Set;

public class Node implements Comparable<Node> {
    public final static Node NULL = new Node(Long.MAX_VALUE);

	Node(long rank) {
		this.rank = rank;
	}

	public static class Target {
	    Target(TransactionHandler<Unit> action, Node node) {
	        this.action = action;
	        this.node = node;
	    }
	    final TransactionHandler<Unit> action;
	    final Node node;
    }

	private long rank;
    List<Target> listeners = new ArrayList<>();

	/**
	 * @return true if any changes were made. 
	 */
	boolean linkTo(TransactionHandler<Unit> action, Node target) {
		boolean changed = target.ensureBiggerThan(rank, new HashSet<Node>());
		listeners.add(new Target(action, target));
		return changed;
	}

	void unlinkTo(Node target) {
	    for (int i = 0; i < listeners.size(); i++)
            if (listeners.get(i).node == target) {
                listeners.remove(i);
                break;
            }
	}

	private boolean ensureBiggerThan(long limit, Set<Node> visited) {
		if (rank > limit || visited.contains(this))
			return false;

		visited.add(this);
		rank = limit + 1;
		for (Target l : listeners)
			l.node.ensureBiggerThan(rank, visited);
		return true;
	}

	@Override
	public int compareTo(Node o) {
		if (rank < o.rank) return -1;
		if (rank > o.rank) return 1;
		return 0;
	}
}
